applications/GeoMechanicsApplication/custom_constitutive/mohr_coulomb_with_tension_cutoff.cpp

```diff
@@ -0,0 +1,272 @@
+        if (derivative_flow_function <= 0.0) {
+            principal_trial_stress_vector =
+                this->ReturnStressAtCornerReturnZone(principal_trial_stress_vector, corner_point);
+            StressStrainUtilities::ReorderEigenValuesAndVectors(principal_trial_stress_vector, rotation_matrix);
+            mStressVector = StressStrainUtilities::RotatePrincipalStresses(
+                principal_trial_stress_vector, rotation_matrix, mpConstitutiveDimension->GetStrainSize());
+            continue;
+        }
+
+        // Regular failure region
+        principal_trial_stress_vector = this->ReturnStressAtRegularFailureZone(
+            principal_trial_stress_vector, mCoulombYieldSurface, friction_angle_in_rad, cohesion);
+        StressStrainUtilities::ReorderEigenValuesAndVectors(principal_trial_stress_vector, rotation_matrix);
+        mStressVector = StressStrainUtilities::RotatePrincipalStresses(
+            principal_trial_stress_vector, rotation_matrix, mpConstitutiveDimension->GetStrainSize());
+    }
+}
+
+Vector MohrCoulombWithTensionCutOff::ReturnStressAtAxialZone(const Vector& rPrincipalTrialStressVector,
+                                                             double TensileStrength) const
+{
+    Vector result = rPrincipalTrialStressVector;
+    result(0)     = TensileStrength;
+    result(2) = TensileStrength - rPrincipalTrialStressVector(0) + rPrincipalTrialStressVector(2);
+    return result;
+}
+
+Vector MohrCoulombWithTensionCutOff::ReturnStressAtCornerReturnZone(const Vector& rPrincipalTrialStressVector,
+                                                                    const Vector& rCornerPoint) const
+{
+    Vector result = rPrincipalTrialStressVector;
+    result(0)     = rCornerPoint(0) + rCornerPoint(1);
+    result(2)     = rCornerPoint(0) - rCornerPoint(1);
+    return result;
+}
+
+Vector MohrCoulombWithTensionCutOff::ReturnStressAtRegularFailureZone(const Vector& rPrincipalTrialStressVector,
+                                                                      const CoulombYieldSurface& rCoulombYieldSurface,
+                                                                      double FrictionAngle,
+                                                                      double Cohesion) const
+{
+    Vector result = rPrincipalTrialStressVector;
+    const auto flowDerivative = rCoulombYieldSurface.DerivateOfFlowFunction(rPrincipalTrialStressVector);
+    const auto cof1 = (1.0 + std::sin(FrictionAngle)) / (1.0 - std::sin(FrictionAngle));
+    const auto cof2 = 2.0 * Cohesion * std::cos(FrictionAngle) / (1.0 - std::sin(FrictionAngle));
+    const auto det  = cof1 * flowDerivative(0) - flowDerivative(2);
+    const auto lambda = (rPrincipalTrialStressVector(2) + cof2 - rPrincipalTrialStressVector(0) * cof1) / det;
+    result(0) = lambda * flowDerivative(0) + rPrincipalTrialStressVector(0);
+    result(2) = result(0) * cof1 - cof2;
+    return result;
```

Comment:
We can simplify this as follows:
```suggestion
    return rPrincipalTrialStressVector + lambda * flowDerivative;
```

You can then also remove the variable `result`.